builder/builder.go

```diff
@@ -110,18 +110,49 @@ func (b *Builder) MakeImageAsTarBuilderFun() states.DockerBuilderFun {
 	}
 }
 
+type successPrinter struct {
+	printedOnce []sync.Once
+	printFunc   []func()
+	printIndex  int
+}
+
+func NewSuccessPrinter(funcs ...func()) *successPrinter {
+	printedOnce := []sync.Once{}
+	for i := 0; i < len(funcs); i++ {
+		printedOnce = append(printedOnce, sync.Once{})
+	}
+
+	return &successPrinter{printedOnce, funcs, 0}
+}
+
+func (sp *successPrinter) PrintCurrentSuccess() {
+	sp.printedOnce[sp.printIndex].Do(sp.printFunc[sp.printIndex])
+}
+
+func (sp *successPrinter) IncrementIndex() {
+	sp.printIndex++
+}
+
 func (b *Builder) convertAndBuild(ctx context.Context, target domain.Target, opt BuildOpt) (*states.MultiTarget, error) {
 	outDir, err := ioutil.TempDir(".", ".tmp-earthly-out")
 	if err != nil {
 		return nil, errors.Wrap(err, "mk temp dir for artifacts")
 	}
 	defer os.RemoveAll(outDir)
-	var successOnce sync.Once
-	successFun := func() {
-		if opt.PrintSuccess {
-			b.s.sm.SetSuccess()
-		}
-	}
+
+	sp := NewSuccessPrinter(
+		func() {
+			if opt.PrintSuccess {
+				b.s.sm.SetSuccess("")
+			}
+		},
+		func() {
+			if opt.PrintSuccess {
+				b.s.sm.SetSuccess("--push")
+			}
+		},
+	)
+
```

Comment:
Im not super hot here on this seeing it in the code review window; I could go to a factory like I had before, but still pump up the array...